File: python/dq_finance/registry.py
"""The gate.

Nothing reaches a calculation without passing through here, and this refuses to
hand back a value that is unverified or outside its effective window. That is the
no-assumed-values doctrine expressed as control flow rather than as a paragraph
someone might remember.

A blank cell is safer than a wrong cell — so the failure mode is a raised
exception, never a fallback, a default, or a stale value.

This mirrors ``src/registry.ts`` exactly. ``data/conformance.json`` is executed by
both test suites and proves the two agree.
"""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, Mapping

_DATA_DIR = Path(__file__).resolve().parents[2] / "data"
_ISO_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


class UnverifiedConstantError(Exception):
    """Raised when a constant has not been checked against its source by a human."""

    def __init__(self, name: str, entry: Mapping[str, Any]) -> None:
        source = entry["source"]
        super().__init__(
            f'"{name}" has not been verified against its source and cannot be used.\n'
            f'  Authority: {source["authority"]}\n'
            f'  Citation:  {source["citation"]}\n'
            f'  Source:    {source["url"]}\n\n'
            "Open the URL, confirm the figure, then set status to \"verified\" with "
            "verifiedBy and verifiedOn. Do not work around this by inlining the number."
        )


class ConstantOutOfEffectError(Exception):
    """Raised when a constant is read for a date outside its effective window."""

    def __init__(self, name: str, entry: Mapping[str, Any], as_of: str) -> None:
        super().__init__(
            f'"{name}" is not in effect on {as_of} '
            f'(effective {entry["effectiveFrom"]} to {entry.get("effectiveTo") or "open"}).\n'
            f'  Source: {entry["source"]["url"]}\n\n'
            "A new tax year needs a new entry, verified against the current publication. "
            "Do not extend effectiveTo without re-checking the source — indexation moves these."
        )


class MalformedConstantError(Exception):
    """Raised when an entry itself is broken. A defect in dq-finance, not the caller."""

    def __init__(self, name: str, problem: str) -> None:
        super().__init__(
            f'"{name}" is malformed: {problem}. This is a defect in dq-finance, not in the caller.'
        )


class UnknownConstantError(Exception):
    """Raised when a name is not in the registry."""

    def __init__(self, name: str) -> None:
        super().__init__(
            f'"{name}" is not in the registry. Add it to data/constants.json with a primary source.'
        )


def _load(filename: str) -> dict[str, Any]:
    return json.loads((_DATA_DIR / filename).read_text(encoding="utf-8"))


#: Every sourced entry, keyed by a stable dotted name. Loaded from the shared JSON
#: that the TypeScript reader also loads, so the two cannot drift.
ALL_ENTRIES: dict[str, Any] = _load("constants.json")


def is_schedule(entry: Mapping[str, Any]) -> bool:
    return "tiers" in entry


def validate(name: str, entry: Mapping[str, Any]) -> None:
    """Structural check. Runs on every read and across the registry in tests."""
    source = entry.get("source") or {}
    if not str(source.get("authority", "")).strip():
        raise MalformedConstantError(name, "source.authority is empty")
    if not str(source.get("citation", "")).strip():
        raise MalformedConstantError(name, "source.citation is empty")
    url = str(source.get("url", "")).strip()
    if not url:
        raise MalformedConstantError(name, "source.url is empty")
    if not url.startswith("https://"):
        raise MalformedConstantError(name, f'source.url must be https, got "{url}"')
    if not _ISO_DATE.match(str(source.get("retrieved", ""))):
        raise MalformedConstantError(
            name, f'source.retrieved must be YYYY-MM-DD, got "{source.get("retrieved")}"'
        )

    effective_from = str(entry.get("effectiveFrom", ""))
    effective_to = entry.get("effectiveTo")
    if not _ISO_DATE.match(effective_from):
        raise MalformedConstantError(name, f'effectiveFrom must be YYYY-MM-DD, got "{effective_from}"')
    if effective_to is not None and not _ISO_DATE.match(str(effective_to)):
        raise MalformedConstantError(name, f'effectiveTo must be YYYY-MM-DD or null, got "{effective_to}"')
    if effective_to is not None and str(effective_to) < effective_from:
        raise MalformedConstantError(name, "effectiveTo is before effectiveFrom")

    if entry.get("status") == "verified":
        if not str(entry.get("verifiedBy", "")).strip():
            raise MalformedConstantError(name, 'status is "verified" but verifiedBy is empty')
        if not _ISO_DATE.match(str(entry.get("verifiedOn", ""))):
            raise MalformedConstantError(
                name, 'status is "verified" but verifiedOn is not a YYYY-MM-DD date'
            )

    if is_schedule(entry):
        tiers = entry["tiers"]
        if not tiers:
            raise MalformedConstantError(name, "schedule has no tiers")
        previous_to: int | None = None
        for i, tier in enumerate(tiers):
            frm, to, rate = tier["from"], tier.get("to"), tier["rate"]
            # bool is a subclass of int in Python; exclude it explicitly.
            if not isinstance(frm, int) or isinstance(frm, bool):
                raise MalformedConstantError(name, f'tier {i} "from" must be an integer (cents), got {frm}')
            if to is not None and (not isinstance(to, int) or isinstance(to, bool)):
                raise MalformedConstantError(name, f'tier {i} "to" must be an integer (cents) or null')
            if to is not None and to <= frm:
                raise MalformedConstantError(name, f"tier {i} is empty or inverted")
            if i > 0 and frm != previous_to:
                raise MalformedConstantError(
                    name,
                    f"tier {i} starts at {frm} but tier {i - 1} ends at {previous_to} "
                    "— schedules must be contiguous",
                )
            if rate < 0 or rate > 1:
                raise MalformedConstantError(
                    name, f"tier {i} rate {rate} is not a decimal fraction between 0 and 1"
                )
            previous_to = to
        if previous_to is not None:
            raise MalformedConstantError(name, "the final tier must be open-ended (to: null)")
        return

    value = entry.get("value")
    unit = entry.get("unit")
    if not isinstance(value, (int, float)) or isinstance(value, bool) or value != value:
        raise MalformedConstantError(name, "value is not a finite number")
    if unit == "cad_cents" and (not isinstance(value, int) or isinstance(value, bool)):
        raise MalformedConstantError(
            name, f"cad_cents value {value} is not an integer — money is never a float"
        )
    if unit == "rate_decimal" and (value < 0 or value > 1):
        raise MalformedConstantError(
            name, f"rate_decimal value {value} is outside 0..1 — 6.25% is 0.0625, not 6.25"
        )


def in_effect(entry: Mapping[str, Any], as_of: str) -> bool:
    if as_of < entry["effectiveFrom"]:
        return False
    effective_to = entry.get("effectiveTo")
    if effective_to is not None and as_of > effective_to:
        return False
    return True


def is_usable(entry: Mapping[str, Any], as_of: str) -> bool:
    """Non-throwing check, for building a verification queue or a status page."""
    return entry.get("status") == "verified" and in_effect(entry, as_of)


def resolve_from_registry(
    registry: Mapping[str, Any], name: str, as_of: str
) -> int | float | list[dict[str, Any]]:
    if not _ISO_DATE.match(as_of):
        raise MalformedConstantError(name, f'asOf must be a YYYY-MM-DD date, got "{as_of}"')
    entry = registry.get(name)
    if entry is None:
        raise UnknownConstantError(name)
    validate(name, entry)
    if entry.get("status") != "verified":
        raise UnverifiedConstantError(name, entry)
    if not in_effect(entry, as_of):
        raise ConstantOutOfEffectError(name, entry, as_of)
    return entry["tiers"] if is_schedule(entry) else entry["value"]


def resolve(name: str, as_of: str) -> int | float | list[dict[str, Any]]:
    """Read a constant by name.

    Raises unless it is well-formed, verified, and in effect on ``as_of``.

    :param as_of: ISO date the calculation applies to — a transaction date or a
        tax-year date, never "today" by default. Passing today's date to a
        historical calculation is how last year's rate silently reaches this
        year's filing, so the caller must be explicit.
    """
    return resolve_from_registry(ALL_ENTRIES, name, as_of)


def get_entry(name: str) -> Mapping[str, Any] | None:
    """The raw entry, including provenance, for inspection and reporting."""
    return ALL_ENTRIES.get(name)


def verification_queue() -> list[dict[str, Any]]:
    """Everything still awaiting a human check."""
    return [
        {
            "name": name,
            "authority": entry["source"]["authority"],
            "url": entry["source"]["url"],
            "note": entry.get("note"),
        }
        for name, entry in ALL_ENTRIES.items()
        if entry.get("status") != "verified"
    ]

File: src/registry.ts
/**
 * The gate.
 *
 * Nothing in this library reaches a calculation without passing through here,
 * and this refuses to hand back a value that is unverified or outside its
 * effective window. That is the no-assumed-values doctrine expressed as
 * control flow rather than as a paragraph someone might remember.
 *
 * A blank cell is safer than a wrong cell — so the failure mode is a thrown
 * error, never a fallback, a default, or a stale value.
 */

import { isSchedule, type Sourced, type SourcedSchedule, type SourcedValue, type Tier } from "./types.ts";

export class UnverifiedConstantError extends Error {
  constructor(name: string, entry: Sourced) {
    super(
      `"${name}" has not been verified against its source and cannot be used.\n` +
        `  Authority: ${entry.source.authority}\n` +
        `  Citation:  ${entry.source.citation}\n` +
        `  Source:    ${entry.source.url}\n\n` +
        `Open the URL, confirm the figure, then set status to "verified" with ` +
        `verifiedBy and verifiedOn. Do not work around this by inlining the number.`,
    );
    this.name = "UnverifiedConstantError";
  }
}

export class ConstantOutOfEffectError extends Error {
  constructor(name: string, entry: Sourced, asOf: string) {
    super(
      `"${name}" is not in effect on ${asOf} ` +
        `(effective ${entry.effectiveFrom} to ${entry.effectiveTo ?? "open"}).\n` +
        `  Source: ${entry.source.url}\n\n` +
        `A new tax year needs a new entry, verified against the current publication. ` +
        `Do not extend effectiveTo without re-checking the source — indexation moves these.`,
    );
    this.name = "ConstantOutOfEffectError";
  }
}

export class UnknownConstantError extends Error {
  constructor(name: string) {
    super(`"${name}" is not in the registry. Add it to data/constants.json with a primary source.`);
    this.name = "UnknownConstantError";
  }
}

export class MalformedConstantError extends Error {
  constructor(name: string, problem: string) {
    super(`"${name}" is malformed: ${problem}. This is a defect in dq-finance, not in the caller.`);
    this.name = "MalformedConstantError";
  }
}

const ISO_DATE = /^\d{4}-\d{2}-\d{2}$/;

/**
 * Structural check. Runs on every read and across the whole registry in tests,
 * so a badly formed entry fails loudly rather than being silently trusted.
 */
export function validate(name: string, entry: Sourced): void {
  const { source } = entry;
  if (!source?.authority?.trim()) throw new MalformedConstantError(name, "source.authority is empty");
  if (!source?.citation?.trim()) throw new MalformedConstantError(name, "source.citation is empty");
  if (!source?.url?.trim()) throw new MalformedConstantError(name, "source.url is empty");
  if (!/^https:\/\//.test(source.url)) {
    throw new MalformedConstantError(name, `source.url must be https, got "${source.url}"`);
  }
  if (!ISO_DATE.test(source.retrieved)) {
    throw new MalformedConstantError(name, `source.retrieved must be YYYY-MM-DD, got "${source.retrieved}"`);
  }
  if (!ISO_DATE.test(entry.effectiveFrom)) {
    throw new MalformedConstantError(name, `effectiveFrom must be YYYY-MM-DD, got "${entry.effectiveFrom}"`);
  }
  if (entry.effectiveTo !== null && !ISO_DATE.test(entry.effectiveTo)) {
    throw new MalformedConstantError(name, `effectiveTo must be YYYY-MM-DD or null, got "${entry.effectiveTo}"`);
  }
  if (entry.effectiveTo !== null && entry.effectiveTo < entry.effectiveFrom) {
    throw new MalformedConstantError(name, "effectiveTo is before effectiveFrom");
  }
  if (entry.status === "verified") {
    if (!entry.verifiedBy?.trim()) {
      throw new MalformedConstantError(name, 'status is "verified" but verifiedBy is empty');
    }
    if (!ISO_DATE.test(entry.verifiedOn ?? "")) {
      throw new MalformedConstantError(name, 'status is "verified" but verifiedOn is not a YYYY-MM-DD date');
    }
  }

  if (isSchedule(entry)) {
    if (entry.tiers.length === 0) throw new MalformedConstantError(name, "schedule has no tiers");
    let previousTo: number | null = null;
    for (const [i, tier] of entry.tiers.entries()) {
      if (!Number.isInteger(tier.from)) {
        throw new MalformedConstantError(name, `tier ${i} "from" must be an integer (cents), got ${tier.from}`);
      }
      if (tier.to !== null && !Number.isInteger(tier.to)) {
        throw new MalformedConstantError(name, `tier ${i} "to" must be an integer (cents) or null`);
      }
      if (tier.to !== null && tier.to <= tier.from) {
        throw new MalformedConstantError(name, `tier ${i} is empty or inverted`);
      }
      if (i > 0 && tier.from !== previousTo) {
        throw new MalformedConstantError(
          name,
          `tier ${i} starts at ${tier.from} but tier ${i - 1} ends at ${previousTo} — schedules must be contiguous with no gap or overlap`,
        );
      }
      if (tier.rate < 0 || tier.rate > 1) {
        throw new MalformedConstantError(name, `tier ${i} rate ${tier.rate} is not a decimal fraction between 0 and 1`);
      }
      previousTo = tier.to;
    }
    if (previousTo !== null) {
      throw new MalformedConstantError(name, "the final tier must be open-ended (to: null)");
    }
    return;
  }

  const v = entry as SourcedValue;
  if (!Number.isFinite(v.value)) throw new MalformedConstantError(name, "value is not a finite number");
  if (v.unit === "cad_cents" && !Number.isInteger(v.value)) {
    throw new MalformedConstantError(name, `cad_cents value ${v.value} is not an integer — money is never a float`);
  }
  if (v.unit === "rate_decimal" && (v.value < 0 || v.value > 1)) {
    throw new MalformedConstantError(name, `rate_decimal value ${v.value} is outside 0..1 — 6.25% is 0.0625, not 6.25`);
  }
}

export function inEffect(entry: Sourced, asOf: string): boolean {
  if (asOf < entry.effectiveFrom) return false;
  if (entry.effectiveTo !== null && asOf > entry.effectiveTo) return false;
  return true;
}

/**
 * Read a constant. Throws unless it is well-formed, verified, and in effect.
 *
 * @param asOf ISO date the calculation applies to — a transaction date or a
 *   tax-year date, never "today" by default. Passing today's date to a
 *   historical calculation is how last year's rate silently reaches this year's
 *   filing, so the caller must be explicit.
 */
export function resolveFromRegistry(
  registry: Readonly<Record<string, Sourced>>,
  name: string,
  asOf: string,
): number | readonly Tier[] {
  if (!ISO_DATE.test(asOf)) {
    throw new MalformedConstantError(name, `asOf must be a YYYY-MM-DD date, got "${asOf}"`);
  }
  const entry = registry[name];
  if (!entry) throw new UnknownConstantError(name);
  validate(name, entry);
  if (entry.status !== "verified") throw new UnverifiedConstantError(name, entry);
  if (!inEffect(entry, asOf)) throw new ConstantOutOfEffectError(name, entry, asOf);
  return isSchedule(entry) ? (entry as SourcedSchedule).tiers : (entry as SourcedValue).value;
}

/** Non-throwing check, for building a verification queue or a status page. */
export function isUsable(entry: Sourced, asOf: string): boolean {
  return entry.status === "verified" && inEffect(entry, asOf);
}
